### PR TITLE
MsgsCertificate Test

### PR DESCRIPTION
**Pull request by oded282**
- Created unit tests for 'MsgsCertificare'.
- Repaired 'InConsistent' function in 'MsgsCertificate',
 and tested the desirable consistency behaviour.
- Include needed files.

### CODE REVIEW

File: bftengine/tests/testMsgsCertificate/CMakeLists.txt [hunk 1]
```diff
@@ -3,6 +3,12 @@ find_package(GTest REQUIRED)
 add_executable(msgsCertificate_test msgsCertificate_test.cpp $<TARGET_OBJECTS:logging_dev>)
 add_test(msgsCertificate_test msgsCertificate_test)
 
+target_include_directories(msgsCertificate_test
+        PRIVATE
+        ${bftengine_SOURCE_DIR}/src/preprocessor
```
Comment (teoparvanov) [inline bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: why do you need the preprocessor as a dependency ?

File: bftengine/src/bftengine/messages/MsgsCertificate.hpp [hunk 2]
```diff
@@ -288,7 +288,7 @@ bool MsgsCertificate<T, SelfTrust, SelfIsRequired, KeepAllMsgs, ExternalFunc>::i
 
 template <typename T, bool SelfTrust, bool SelfIsRequired, bool KeepAllMsgs, typename ExternalFunc>
 bool MsgsCertificate<T, SelfTrust, SelfIsRequired, KeepAllMsgs, ExternalFunc>::isInconsistent() const {
-  return (numOfClasses >= maxFails + 1);
+  return (numOfClasses > maxFails + 1);
```
Comment (teoparvanov) [inline bftengine/src/bftengine/messages/MsgsCertificate.hpp:291]: which test case discovered this problem?

@yuliasherman this looks like a missed corner case - what do you think?

Comment (oded282) [reply bftengine/src/bftengine/messages/MsgsCertificate.hpp:291]: - I read the code, and understood the number doesn't fit the protocol.
  Also checked with Guy to be sure, he confirmed that indeed this is incorrect.

Comment (teoparvanov) [reply bftengine/src/bftengine/messages/MsgsCertificate.hpp:291]: Great job! Can you also write a unit test that specifically reproduces this problem?

Comment (teoparvanov) [reply bftengine/src/bftengine/messages/MsgsCertificate.hpp:291]: As discussed offline, the test already exists, and will be mentioned in the PR description.

File: bftengine/tests/testMsgsCertificate/CMakeLists.txt (hunk 1 above)
Comment (oded282) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: In order to use 'ClientReplyMsg' used in my test, I have needed to include '${bftengine_SOURCE_DIR}/src/preprocessor'.

File: bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp [hunk 3]
```diff
@@ -1,10 +1,132 @@
...
+
+  uint8_t reqSeqNum = 0;
+
+  // Create two identical messages and add them.
+
+  ReplicaId replicaId = randomReplicaeId();
+  auto bestMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+  replysCertificate.addMsg(bestMsg, randomReplicaeId());
+
+  replicaId = randomReplicaeId();
```
Comment (teoparvanov) [inline bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp:93]: nit - `randomReplicaeId` -> `randomReplicaId`

File: bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp [hunk 4]
```diff
@@ -1,10 +1,132 @@
...
+      numOfReplicas, fval, numOfRequired, selfReplicaId);
+
+  uint8_t reqSeqNum = 0;
+
+  // Add bunch of new messages
+  for (; reqSeqNum < 3; reqSeqNum++) {
+    ReplicaId replicaId = randomReplicaeId();
+    auto newMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+    replysCertificate.addMsg(newMsg, replicaId);
+  }
```
Comment (HristoStaykov) [inline bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp:41]: I think the "replicaId" shouldn't be drawn from the RNG. If I understand correct we want different replica ID to add in every round.

File: bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp [hunk 5]
```diff
@@ -1,10 +1,132 @@
...
+  replicaId = randomReplicaeId();
+  auto identBestMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+  replysCertificate.addMsg(identBestMsg, randomReplicaeId());
+
+  // Add bunch of new messages
+  for (; reqSeqNum < 4; reqSeqNum++) {
+    replicaId = randomReplicaeId();
+    auto newMsg = new ClientReplyMsg(replicaId, reqSeqNum, reply, replyLen);
+    replysCertificate.addMsg(newMsg, replicaId);
+  }
```
Comment (HristoStaykov) [inline bftengine/tests/testMsgsCertificate/msgsCertificate_test.cpp:102]: Same thing here, if we draw from RNG, it could happen that some replicaId-s will coincide.

File: bftengine/tests/testMsgsCertificate/CMakeLists.txt (hunk 1 above)
Comment (teoparvanov) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: can't you use another message, which is not in the pre-processor?

Comment (oded282) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: Correction:
It was in order to use 'SimpleClientImp' and the method implemented there.

Comment (teoparvanov) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: I don't get it... SimpleClientImp is not defined in the preprocessor.

Comment (oded282) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: SimpleClientImp include "messages/ClientPreProcessRequestMsg.hpp".
My class use SimpleClientImp and this dir is unknown for my class, that's why I added it.

Comment (teoparvanov) [reply bftengine/tests/testMsgsCertificate/CMakeLists.txt:8]: Alright, I see - thanks for clarifying.
